// @sanity/color-input/lib/ColorPicker.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _common = require("react-color/lib/components/common");
var _ui = require("@sanity/ui");
var _icons = require("@sanity/icons");
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _ColorPickerFields = require("./ColorPickerFields");
var _templateObject, _templateObject2;
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }
var ColorBox = (0, _styledComponents.default)(_ui.Box)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  position: absolute;\n  top: 0;\n  left: 0;\n  width: 100%;\n  height: 100%;\n"])));
var ReadOnlyContainer = (0, _styledComponents.default)(_ui.Flex)(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n  margin-top: 6rem;\n  background-color: var(--card-bg-color);\n  position: relative;\n  width: 100%;\n"])));
var ColorPicker = _ref => {
  var width = _ref.width,
    rgb = _ref.rgb,
    hex = _ref.hex,
    hsv = _ref.hsv,
    hsl = _ref.hsl,
    onChange = _ref.onChange,
    onUnset = _ref.onUnset,
    disableAlpha = _ref.disableAlpha,
    renderers = _ref.renderers,
    readOnly = _ref.readOnly;
  return /*#__PURE__*/_react.default.createElement("div", {
    style: {
      width
    }
  }, /*#__PURE__*/_react.default.createElement(_ui.Card, {
    padding: 1,
    border: true,
    radius: 1
  }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    space: 2
  }, !readOnly && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_ui.Card, {
    overflow: "hidden",
    style: {
      position: 'relative',
      height: '5em'
    }
  }, /*#__PURE__*/_react.default.createElement(_common.Saturation, {
    is: "Saturation",
    onChange: onChange,
    hsl: hsl,
    hsv: hsv
  })), /*#__PURE__*/_react.default.createElement(_ui.Card, {
    shadow: 1,
    radius: 3,
    overflow: "hidden",
    style: {
      position: 'relative',
      height: '10px'
    }
  }, /*#__PURE__*/_react.default.createElement(_common.Hue, {
    is: "Hue",
    hsl: hsl,
    onChange: !readOnly && onChange
  })), !disableAlpha && /*#__PURE__*/_react.default.createElement(_ui.Card, {
    shadow: 1,
    radius: 3,
    overflow: "hidden",
    style: {
      position: 'relative',
      height: '10px'
    }
  }, /*#__PURE__*/_react.default.createElement(_common.Alpha, {
    is: "Alpha",
    rgb: rgb,
    hsl: hsl,
    renderers: renderers,
    onChange: onChange
  }))), /*#__PURE__*/_react.default.createElement(_ui.Flex, null, /*#__PURE__*/_react.default.createElement(_ui.Card, {
    flex: 1,
    radius: 2,
    overflow: "hidden",
    style: {
      position: 'relative',
      minWidth: '4em'
    }
  }, /*#__PURE__*/_react.default.createElement(_common.Checkboard, null), /*#__PURE__*/_react.default.createElement(ColorBox, {
    style: {
      backgroundColor: "rgba(".concat(rgb.r, ",").concat(rgb.g, ",").concat(rgb.b, ",").concat(rgb.a, ")")
    }
  }), readOnly && /*#__PURE__*/_react.default.createElement(ReadOnlyContainer, {
    padding: 2,
    paddingBottom: 1,
    sizing: "border",
    justify: "space-between"
  }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    space: 3,
    marginTop: 1
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 3,
    weight: "bold"
  }, hex), /*#__PURE__*/_react.default.createElement(_ui.Inline, {
    space: 3
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 1
  }, /*#__PURE__*/_react.default.createElement("strong", null, "RGB: "), rgb.r, " ", rgb.g, " ", rgb.b), /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 1
  }, /*#__PURE__*/_react.default.createElement("strong", null, "HSL: "), " ", Math.round(hsl.h), " ", Math.round(hsl.s), "%", ' ', Math.round(hsl.l)))))), !readOnly && /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "flex-start",
    marginLeft: 2
  }, /*#__PURE__*/_react.default.createElement(_ui.Box, {
    style: {
      width: 200
    }
  }, /*#__PURE__*/_react.default.createElement(_ColorPickerFields.ColorPickerFields, {
    rgb: rgb,
    hsl: hsl,
    hex: hex,
    onChange: onChange,
    disableAlpha: disableAlpha
  })), /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginLeft: 2
  }, /*#__PURE__*/_react.default.createElement(_ui.Button, {
    onClick: onUnset,
    title: "Delete color",
    icon: _icons.TrashIcon,
    tone: "critical"
  })))))));
};
ColorPicker.propTypes = {
  width: _propTypes.default.string,
  hex: _propTypes.default.string,
  hsl: _propTypes.default.object,
  hsv: _propTypes.default.object,
  rgb: _propTypes.default.object,
  onChange: _propTypes.default.func,
  disableAlpha: _propTypes.default.bool,
  readOnly: _propTypes.default.bool,
  renderers: _propTypes.default.func,
  onUnset: _propTypes.default.func
};
ColorPicker.defaultProps = {
  disableAlpha: false
};
var _default = (0, _common.ColorWrap)(ColorPicker);
exports.default = _default;
