// @sanity/color-input/lib/schemas/color.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _ColorInput = _interopRequireDefault(require("../ColorInput"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
/* eslint-disable react/display-name */

var round = val => Math.round(val * 100);
var _default = {
  name: 'color',
  type: 'object',
  title: 'Color',
  inputComponent: _ColorInput.default,
  fields: [{
    title: 'Hex',
    name: 'hex',
    type: 'string'
  }, {
    title: 'Alpha',
    name: 'alpha',
    type: 'number'
  }, {
    title: 'Hue Saturation Lightness',
    name: 'hsl',
    type: 'hslaColor'
  }, {
    title: 'Hue Saturation Value',
    name: 'hsv',
    type: 'hsvaColor'
  }, {
    title: 'Red Green Blue (rgb)',
    name: 'rgb',
    type: 'rgbaColor'
  }],
  preview: {
    select: {
      title: 'hex',
      alpha: 'alpha',
      hex: 'hex',
      hsl: 'hsl'
    },
    prepare(_ref) {
      var title = _ref.title,
        hex = _ref.hex,
        hsl = _ref.hsl,
        alpha = _ref.alpha;
      var subtitle = hex || 'No color set';
      if (hsl) {
        subtitle = "H:".concat(round(hsl.l), " S:").concat(round(hsl.l), " L:").concat(round(hsl.l), " A:").concat(round(alpha));
      }
      return {
        title: title,
        subtitle: subtitle,
        media: () => /*#__PURE__*/_react.default.createElement("div", {
          style: {
            backgroundColor: hex || '#000',
            opacity: alpha || 1,
            position: 'absolute',
            height: '100%',
            width: '100%',
            top: '0',
            left: '0'
          }
        })
      };
    }
  }
};
exports.default = _default;
