// @sanity/preview-kit/dist/_chunks/context-6e668875.cjs
'use strict';

var react = require('react');
function getQueryCacheKey(query, params) {
  return "".concat(query, "-").concat(JSON.stringify(params));
}
function useLoadingListenersContext(ready) {
  const [tick, forceUpdate] = react.useReducer(x => x + 1, 0);
  const mountedRef = react.useRef(true);
  const scheduleUpdate = react.useCallback(() => {
    if (mountedRef.current) {
      react.startTransition(() => forceUpdate());
    }
  }, []);
  react.useEffect(() => {
    mountedRef.current = true;
    scheduleUpdate();
    return () => {
      mountedRef.current = false;
    };
  }, [scheduleUpdate]);
  const loadedListenersContext = react.useMemo(() => tick ? [...ready] : [], [ready, tick]);
  return [loadedListenersContext, scheduleUpdate];
}
const DEFAULT_MAX_DOCUMENTS = 3e3;
const snapshots = /* @__PURE__ */new Map();
const deps = /* @__PURE__ */new Map();
const NoStoreContext = function defineListener(initialSnapshot, query, params) {
  const key = getQueryCacheKey(query, params);
  snapshots.set(key, initialSnapshot);
  if (!deps.has(key)) {
    deps.set(key, 0);
  }
  const subscribe = () => {
    deps.set(key, deps.get(key) + 1);
    return () => {
      deps.set(key, deps.get(key) - 1);
      if (deps.get(key) === 0) {
        snapshots.delete(key);
      }
    };
  };
  const getSnapshot = () => snapshots.has(key) ? snapshots.get(key) : initialSnapshot;
  return {
    subscribe,
    getSnapshot
  };
};
const defineListenerContext = react.createContext(NoStoreContext);
const LoadedListenersContext = react.createContext(null);
exports.DEFAULT_MAX_DOCUMENTS = DEFAULT_MAX_DOCUMENTS;
exports.LoadedListenersContext = LoadedListenersContext;
exports.defineListenerContext = defineListenerContext;
exports.getQueryCacheKey = getQueryCacheKey;
exports.useLoadingListenersContext = useLoadingListenersContext;
