// @vercel/fs-detectors/dist/detectors/local-file-system-detector.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LocalFileSystemDetector = void 0;
const promises_1 = __importDefault(require("fs/promises"));
const path_1 = __importDefault(require("path"));
const filesystem_1 = require("./filesystem");
const error_utils_1 = require("@vercel/error-utils");
class LocalFileSystemDetector extends filesystem_1.DetectorFilesystem {
    constructor(rootPath) {
        super();
        this.rootPath = rootPath;
    }
    async _hasPath(name) {
        try {
            await promises_1.default.stat(this.getFilePath(name));
            return true;
        }
        catch (err) {
            if ((0, error_utils_1.isErrnoException)(err) && err.code === 'ENOENT') {
                return false;
            }
            throw err;
        }
    }
    _readFile(name) {
        return promises_1.default.readFile(this.getFilePath(name));
    }
    async _isFile(name) {
        const stat = await promises_1.default.stat(this.getFilePath(name));
        return stat.isFile();
    }
    async _readdir(name) {
        const dirPath = this.getFilePath(name);
        const dir = await promises_1.default.readdir(dirPath, {
            withFileTypes: true,
        });
        const getType = (dirent) => {
            if (dirent.isFile()) {
                return 'file';
            }
            else if (dirent.isDirectory()) {
                return 'dir';
            }
            else {
                throw new Error(`Dirent was neither file nor directory`);
            }
        };
        return dir.map(dirent => ({
            name: dirent.name,
            path: path_1.default.join(dirPath, dirent.name),
            type: getType(dirent),
        }));
    }
    _chdir(name) {
        return new LocalFileSystemDetector(this.getFilePath(name));
    }
    getFilePath(name) {
        return path_1.default.join(this.rootPath, name);
    }
}
exports.LocalFileSystemDetector = LocalFileSystemDetector;
