// @vercel/fs-detectors/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.packageManagers = exports.isStaticRuntime = exports.isOfficialRuntime = exports.monorepoManagers = exports.getWorkspacePackagePaths = exports.getWorkspaces = exports.workspaceManagers = exports.LocalFileSystemDetector = exports.DetectorFilesystem = exports.getProjectPaths = exports.detectFrameworkVersion = exports.detectFrameworkRecord = exports.detectFramework = exports.detectFileSystemAPI = exports.detectApiExtensions = exports.detectApiDirectory = exports.detectOutputDirectory = exports.detectBuilders = void 0;
var detect_builders_1 = require("./detect-builders");
Object.defineProperty(exports, "detectBuilders", { enumerable: true, get: function () { return detect_builders_1.detectBuilders; } });
Object.defineProperty(exports, "detectOutputDirectory", { enumerable: true, get: function () { return detect_builders_1.detectOutputDirectory; } });
Object.defineProperty(exports, "detectApiDirectory", { enumerable: true, get: function () { return detect_builders_1.detectApiDirectory; } });
Object.defineProperty(exports, "detectApiExtensions", { enumerable: true, get: function () { return detect_builders_1.detectApiExtensions; } });
var detect_file_system_api_1 = require("./detect-file-system-api");
Object.defineProperty(exports, "detectFileSystemAPI", { enumerable: true, get: function () { return detect_file_system_api_1.detectFileSystemAPI; } });
var detect_framework_1 = require("./detect-framework");
Object.defineProperty(exports, "detectFramework", { enumerable: true, get: function () { return detect_framework_1.detectFramework; } });
Object.defineProperty(exports, "detectFrameworkRecord", { enumerable: true, get: function () { return detect_framework_1.detectFrameworkRecord; } });
Object.defineProperty(exports, "detectFrameworkVersion", { enumerable: true, get: function () { return detect_framework_1.detectFrameworkVersion; } });
var get_project_paths_1 = require("./get-project-paths");
Object.defineProperty(exports, "getProjectPaths", { enumerable: true, get: function () { return get_project_paths_1.getProjectPaths; } });
var filesystem_1 = require("./detectors/filesystem");
Object.defineProperty(exports, "DetectorFilesystem", { enumerable: true, get: function () { return filesystem_1.DetectorFilesystem; } });
var local_file_system_detector_1 = require("./detectors/local-file-system-detector");
Object.defineProperty(exports, "LocalFileSystemDetector", { enumerable: true, get: function () { return local_file_system_detector_1.LocalFileSystemDetector; } });
var workspace_managers_1 = require("./workspaces/workspace-managers");
Object.defineProperty(exports, "workspaceManagers", { enumerable: true, get: function () { return workspace_managers_1.workspaceManagers; } });
var get_workspaces_1 = require("./workspaces/get-workspaces");
Object.defineProperty(exports, "getWorkspaces", { enumerable: true, get: function () { return get_workspaces_1.getWorkspaces; } });
var get_workspace_package_paths_1 = require("./workspaces/get-workspace-package-paths");
Object.defineProperty(exports, "getWorkspacePackagePaths", { enumerable: true, get: function () { return get_workspace_package_paths_1.getWorkspacePackagePaths; } });
var monorepo_managers_1 = require("./monorepos/monorepo-managers");
Object.defineProperty(exports, "monorepoManagers", { enumerable: true, get: function () { return monorepo_managers_1.monorepoManagers; } });
var is_official_runtime_1 = require("./is-official-runtime");
Object.defineProperty(exports, "isOfficialRuntime", { enumerable: true, get: function () { return is_official_runtime_1.isOfficialRuntime; } });
Object.defineProperty(exports, "isStaticRuntime", { enumerable: true, get: function () { return is_official_runtime_1.isStaticRuntime; } });
var package_managers_1 = require("./package-managers/package-managers");
Object.defineProperty(exports, "packageManagers", { enumerable: true, get: function () { return package_managers_1.packageManagers; } });
__exportStar(require("./monorepos/get-monorepo-default-settings"), exports);
